### PR TITLE
Fix WebGPU recurrent state graph capture

### PR DESCRIPTION
**Pull request by qjia7**
## Summary

- enable recurrent-state graph-capture handling for WebGPU hybrid decoder models
- keep WebGPU's required separate past and present recurrent buffers and capture one graph variant for each buffer-swap direction
- select alternating variants in standard decoder-only and multimodal decoder states
- restore the standard decoder's canonical recurrent-buffer direction and matching graph variant on a full rewind
- leave pipeline decoder states excluded because their stage runner does not select graph variants
- add deterministic multimodal and text-only Qwen3.5 recurrent-state fixtures and an isolated WebGPU integration test that compares eager and graph-captured execution

## Why

WebGPU cannot alias recurrent past and present buffers through the shared allocator used by CUDA. The buffers therefore swap after each token. A captured graph has stable buffer bindings, so reusing one graph for both swap directions reads stale recurrent state. Capturing and selecting one variant per direction preserves the required WebGPU double buffering.

A full recurrent-state rewind on the standard decoder also needs to restore the buffer orientation and variant together. Resetting only the variant after an odd number of swaps would select a graph captured with the opposite input/output bindings.

## Scope audit

The final diff was compared file-by-file with current `origin/main`. It retains only the recurrent graph-variant implementation, the constructor choices for standard/multimodal/pipeline decoders, reproducible fixtures, and focused regression coverage. The temporary multimodal full-rewind implementation was removed because multimodal rewind is not supported on `main` and is outside this PR's graph-capture scope.

## Validation

- Release C++ `onnxruntime-genai` target builds successfully
- isolated real-WebGPU integration test matches eager and captured recurrent-state values for both multimodal and decoder-only states, including standard-decoder rewind and replay through `0, 1, 2, 3`
- multimodal coverage uses the configured vision-start/image token sequence and `int32` input IDs
- Ruff and Python byte-compilation pass for the modified Python files
- generated decoder fixtures reproduce byte-for-byte and pass ONNX checker
- clang-format and `git diff --check` pass

### CODE REVIEW

File: src/models/io/recurrent_state.cpp [hunk 1]
```diff
@@ -118,18 +118,17 @@ RecurrentState::RecurrentState(State& state)
...
+  // separate past/present buffers are required (always on WebGPU, or by the CUDA diagnostic
+  // override), capture one graph for each direction of the per-step buffer swap.
   bool share_under_graph_capture = true;
   GetEnv("ORTGENAI_SHARE_RECURRENT_STATE_UNDER_GRAPH_CAPTURE", share_under_graph_capture);
-  const bool graph_capture_enabled = !is_webgpu && state_.params_->use_graph_capture;
+  const bool graph_capture_enabled = state_.params_->use_graph_capture;
   share_buffers_ = !is_webgpu &&
                    (graph_capture_enabled ? share_under_graph_capture : share_buffers_configured);
-  graph_double_buffer_ = graph_capture_enabled && !share_buffers_;
+  graph_double_buffer_ = graph_capture_variants_supported && graph_capture_enabled && !share_buffers_;
```
[the remark this answers is not in the document]
Comment (qjia7) [reply src/models/io/recurrent_state.cpp:131]: Fixed in b1c78705. The full-reset path now restores the canonical allocation direction when the current variant is odd, resets graph_buffer_variant_ to 0, and then rebinds the inputs/outputs. Restoring the pointer direction together with the variant is important; resetting only the integer would select a graph captured with the opposite bindings.

**Comment by qjia7** [comment]
Addressed the two suppressed test findings in b1c78705:

- `test/models/qwen3-5` is already checked into the repository, so no download or fixture-generation wiring is needed. I removed the existence guard so a missing fixture now fails instead of silently skipping the regression.
- Both eager and captured model instances now use `validationMode=disabled`, because this is a process-wide WebGPU context option.

**Comment by qjia7** [comment]
Addressed the latest suppressed review findings in 11f39956. The WebGPU regression now performs three forwards (leaving the double buffers in the odd/noncanonical direction), calls 
ewind_to(0), re-appends the prompt on the same generator, and verifies recurrent-state progression through both capture variants and replay. This exposed that Qwen3.5's MultiModalPipelineState did not forward rewind to its decoder state, so the commit adds that forwarding and resets position, KV, and recurrent state consistently. The rebuilt real WebGPU run passes in eager and graph-captured modes: before rewind [0, 1, 2], after rewind [0, 1, 2, 3]. The suggested missing-model guard was not added because 	est/models/qwen3-5 is tracked in both origin/main and this branch; silently skipping it would weaken CI coverage.

**Comment by qjia7** [comment]
Addressed the latest review findings in 8e7e70c2:

- full multimodal rewind now restores prompt mode and recreates/rebinds cached vision, speech, and embedding state;
- the graph-capture regression runs in an isolated process so WebGPU `validationMode=disabled` is effective before context initialization;
- a caller model root without `qwen3-5` falls back to the tracked deterministic fixture, retaining mandatory coverage;
- added a text-only Qwen3.5 decoder fixture and exercised `DecoderOnly_State` as well as `MultiModalPipelineState`, closing the suppressed standard-decoder coverage finding.

The rebuilt real WebGPU regression passes eager and graph-captured execution for both state implementations, including odd-buffer rewind and replay through `[0, 1, 2, 3]`. The C++ target builds, both Python files are Ruff-clean, both fixtures pass ONNX checker, and `git diff --check` passes.

File: src/models/multi_modal.cpp [hunk 2]
```diff
@@ -941,6 +949,10 @@ DeviceSpan<float> MultiModalPipelineState::Run(int current_length, DeviceSpan<in
   return decoder_state_->Run(current_length, next_tokens, next_indices);
 }
 
+void MultiModalPipelineState::RewindTo(size_t index) {
+  decoder_state_->RewindTo(index);
+}
```
[the remark this answers is not in the document]
Comment (qjia7) [reply src/models/multi_modal.cpp]: Fixed in 8e7e70c2. A full rewind now caches and rebinds the original extra inputs, restores prompt mode, and recreates the vision, speech, and embedding states before the prompt is run again. The WebGPU regression now uses actual image inputs and verifies the same generator through rewind and reprefill.

File: test/python/models/test_qwen_fara_models.py [hunk 3]
```diff
@@ -769,6 +769,52 @@ def test_qwen3_5_hybrid_text_generation_webgpu(test_data_path):
     assert generator is not None
 
 
+@pytest.mark.graph_capture
+@pytest.mark.skipif(not _webgpu_plugin_registered, reason="onnxruntime-ep-webgpu plugin not installed")
+def test_qwen3_5_hybrid_graph_capture_advances_recurrent_state_webgpu(test_data_path):
+    """Graph capture must preserve both directions of WebGPU recurrent-state double buffering."""
+    model_path = os.fspath(Path(test_data_path) / "qwen3-5")
```
[the remark this answers is not in the document]
Comment (qjia7) [reply test/python/models/test_qwen_fara_models.py]: Fixed in 8e7e70c2 with stronger coverage than a skip. If the caller-supplied model root does not contain qwen3-5, the test falls back to the deterministic fixture tracked under test/models/qwen3-5. It asserts if that tracked fixture is unavailable, so the regression cannot silently stop running.

File: test/python/models/test_qwen_fara_models.py [hunk 4]
```diff
@@ -769,6 +769,52 @@ def test_qwen3_5_hybrid_text_generation_webgpu(test_data_path):
...
+
+    def run(enable_graph_capture):
+        config = og.Config(model_path)
+        config.clear_providers()
+        config.append_provider("webgpu")
+        config.set_provider_option("webgpu", "enableGraphCapture", "1" if enable_graph_capture else "0")
+        # WebGPU context options are process-wide, so both model instances must use the same mode.
+        config.set_provider_option("webgpu", "validationMode", "disabled")
```
[the remark this answers is not in the document]
Comment (qjia7) [reply test/python/models/test_qwen_fara_models.py:841]: Fixed in 8e7e70c2. The graph-capture test now launches its test body in a fresh subprocess, so validationMode=disabled is set before that process creates any WebGPU model or initializes the process-wide context.

File: src/models/multi_modal.cpp [hunk 5]
```diff
@@ -716,6 +716,14 @@ DeviceSpan<float> DecoderState::Run(int current_length, DeviceSpan<int32_t>& nex
   return logits_.Get();
 }
 
+void DecoderState::RewindTo(size_t index) {
+  position_inputs_->RewindTo(index);
+  if (kv_cache_)
+    kv_cache_->RewindTo(index);
+  if (recurrent_state_)
+    recurrent_state_->RewindTo(index);
```
[the remark this answers is not in the document]
Comment (qjia7) [reply src/models/multi_modal.cpp]: Resolved in 13987bf3 by narrowing the PR to its graph-capture scope. The newly added multimodal full-rewind path was removed, so this PR no longer sends Qwen-VL position state through an unsupported full rewind. Rewind/variant restoration remains covered on the standard DecoderOnly_State, which already supports rewind on main.

File: test/python/models/test_qwen_fara_models.py [hunk 6]
```diff
@@ -769,6 +773,123 @@ def test_qwen3_5_hybrid_text_generation_webgpu(test_data_path):
...
+        model = og.Model(config)
+        params = og.GeneratorParams(model)
+        params.set_search_options(do_sample=False, max_length=8)
+        generator = og.Generator(model, params)
+        prompt = [10, 20, 30, 40]
+        if multimodal:
+            inputs = og.NamedTensors()
+            inputs["input_ids"] = np.asarray([prompt], dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (qjia7) [reply test/python/models/test_qwen_fara_models.py]: Fixed in 13987bf3. The multimodal input_ids tensor now uses np.int32, matching Generator::SetInputs. The prompt also contains the configured vision_start_token_id and image_token_id so the test executes Qwen multimodal 3-D position initialization.

**Comment by qjia7** [comment]
Addressed the latest Copilot findings and completed a full scope audit in 13987bf3:

- changed multimodal `input_ids` to `np.int32`;
- changed the prompt to include `vision_start_token_id=151652` followed by `image_token_id=151655`, exercising Qwen multimodal 3-D position initialization;
- removed the branch's newly introduced multimodal full-rewind implementation after comparing with `main`; multimodal rewind is not supported there and was outside this graph-capture fix;
- retained full-rewind coverage on `DecoderOnly_State`, whose existing rewind API validates restoration of the recurrent buffer direction and graph variant;
- compared every remaining file with current `origin/main`; no equivalent WebGPU graph-variant enablement exists there, and every remaining source/fixture/test change is required for the fix or its regression coverage.

Validation after the cleanup: Release C++ build passed, the isolated real-WebGPU eager/captured test passed for multimodal and text-only states, Ruff and Python compilation passed, decoder fixtures reproduced byte-for-byte and passed ONNX checker, and clang-format plus `git diff --check` passed.

File: test/python/models/test_qwen_fara_models.py [hunk 7]
```diff
@@ -769,6 +773,127 @@ def test_qwen3_5_hybrid_text_generation_webgpu(test_data_path):
...
+            env={_QWEN35_WEBGPU_GRAPH_CAPTURE_CHILD: "1"},
+            log=log,
+        )
+        return
+
+    model_path = model_root / "qwen3-5"
+    if not model_path.is_dir():
+        # The deterministic fixture is tracked in the repository, so retain coverage when a
+        # standalone caller supplies a different model root instead of silently skipping.
+        model_path = tracked_model_root / "qwen3-5"
```
[the remark this answers is not in the document]
Comment (qjia7) [reply test/python/models/test_qwen_fara_models.py]: Fixed in 29332e3d. This regression now always uses the deterministic qwen3-5 fixture tracked in the repository, so a caller-supplied directory can no longer select an incomplete or older fixture.

File: test/python/models/test_qwen_fara_models.py [hunk 8]
```diff
@@ -769,6 +773,127 @@ def test_qwen3_5_hybrid_text_generation_webgpu(test_data_path):
...
+        config.clear_providers()
+        config.append_provider("webgpu")
+        config.set_provider_option("webgpu", "enableGraphCapture", "1" if enable_graph_capture else "0")
+        # WebGPU context options are process-wide, so both model instances must use the same mode.
+        config.set_provider_option("webgpu", "validationMode", "disabled")
+        model = og.Model(config)
+        params = og.GeneratorParams(model)
+        params.set_search_options(do_sample=False, max_length=8)
+        generator = og.Generator(model, params)
+        prompt = [151652, 151655, 10, 20]
```
[the remark this answers is not in the document]
Comment (qjia7) [reply test/python/models/test_qwen_fara_models.py]: Fixed in 29332e3d. The multimodal case uses the configured vision-start/image-token sequence, while the qwen3_5_text case now uses a separate text-only prompt and does not require image_grid_thw.

File: test/python/models/test_qwen_fara_models.py [hunk 9]
```diff
@@ -769,6 +773,127 @@ def test_qwen3_5_hybrid_text_generation_webgpu(test_data_path):
...
+            generator.set_inputs(inputs)
+        else:
+            generator.append_tokens(prompt)
+
+        # Exercise each direction of the recurrent-state double buffers. For the standard decoder,
+        # stop in the noncanonical direction and also verify its existing full-rewind support.
+        state_values_before_rewind = [float(np.asarray(generator.get_logits()).reshape(-1)[0])]
+        for _ in range(2 if test_rewind else 3):
+            generator.generate_next_token()
+            state_values_before_rewind.append(float(np.asarray(generator.get_logits()).reshape(-1)[0]))
```
[the remark this answers is not in the document]
Comment (qjia7) [reply test/python/models/test_qwen_fara_models.py]: Addressed in 29332e3d by consuming generator.get_logits() directly and removing the redundant np.asarray wrappers in both forward and replay blocks. In this Python API, PyGenerator::GetLogits (src/python/python.cpp) returns pybind11::array_t<float>, so as_numpy() is not available or needed. The focused real-WebGPU test passes with the direct NumPy reshape.

**Comment by qjia7** [comment]
Addressed the latest Copilot review in 29332e3d:

- the regression always selects the tracked deterministic `test/models/qwen3-5` fixture, so an incomplete caller model directory cannot shadow it;
- multimodal and text-only runs now use separate prompts, avoiding vision tokens without grid inputs in `qwen3_5_text`;
- removed redundant `np.asarray` conversions and read logits directly from `Generator.get_logits()`, which this Python binding returns as `pybind11::array_t<float>`/NumPy rather than `OgaTensor`.

The isolated real-WebGPU regression passes for multimodal and text-only eager/captured paths, including text-only rewind. Ruff, Python compilation, formatting, and `git diff --check` also pass.

**Review by prathikr — APPROVED** [review]

**Review by baijumeswani — APPROVED** [review]
